_posts/2019-08-18-Securing-Web-Applications-With-Keycloak.md

```diff
@@ -0,0 +1,608 @@
+* [Implementing Security](#implementing-security)
+  * [Implicit Flow versus Code Flow + PKCE](#implicit-flow-versus-code-flow--pkce)
+  * [JSON Web Token (JWT)](#json-web-token-jwt)
+  * [Resource Server in Spring Boot](#resource-server-in-spring-boot)
+  * [Securing The Angular application](#securing-the-angular-application)
+
+# Why Keycloak as Authentication Server
+You can find several platforms that handle user logins and resource access management such as Keycloak, OKTA, OpenAM, etc. 
+All those platforms have their own features and possibilities that may be useful for your use case. 
+In this article, we choose [Keycloak](https://www.keycloak.org/) as [authentication](https://en.wikipedia.org/wiki/Authentication) and [authorization](https://en.wikipedia.org/wiki/Authorization) server which is an open-source identity and access management platform (IAM) from Red Hat's Jboss. 
+We have chosen for Keycloak because it is [open-source](https://github.com/keycloak/keycloak) and [well-documented](https://www.keycloak.org/documentation.html).
+
+Keycloak comes with several handy features build-in:
+* Two-factor authentication
+* Bruteforce detection
+* Social login (Facebook, Twitter, Google…)
+* LDAP/AD integration
+* …
+
+We will go over the basics to get you started.
+
+# Setting up a Keycloak Server
+
+Keycloak supports [multiple ways](https://www.keycloak.org/docs/latest/server_installation/index.html#_operating-mode) to be set up.
+For non-production purposes, it's easiest to just [download](https://www.keycloak.org/downloads.html) and run the standalone.
+At the time of writing, the latest release version was `6.0.1`.
+
+```bash
+$ curl https://downloads.jboss.org/keycloak/6.0.1/keycloak-6.0.1.zip --output keycloak-6.0.1.zip
+$ unzip keycloak-6.0.1.zip
+$ cd keycloak-6.0.1/bin/
+$ ./standalone
+```
+
+> For Windows users, there is also a `standalone.bat` in the same folder.
+
+The Keycloak server is now running on port 8080.
+
+> Use `-Djboss.socket.binding.port-offset` to change the port.  
+`-Djboss.socket.binding.port-offset=1000` will run the server on port 8080 + 1000 = 9080 
+
+Go to [http://localhost:8080](http://localhost:8080) and create an administrator account.
+You can now click on `Administration Console >` and log in using the account you've just created.  
+
+You are now on the pre-defined Master realm.
+
+> A realm manages a set of users, credentials, roles, and groups. 
+A user belongs to and logs into a realm and they are isolated from one another and can only manage and authenticate the users that they control.  
+
+The master realm is the highest level in the hierarchy. 
+Admin accounts in this realm have permissions to view and manage any other realm. 
+It's best to create a new realm to manage our application and users.
+
+## Creating a New Realm
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/create_realm.png" width="30%">
```

Comment:
Add an image with eg. `<img class="image fit" src="{{ '/img/devoxx-ma-2018/speakers-dinner.jpg' | prepend: site.baseurl }}" alt="Speakers Dinner" width="60%" />`, otherwise this won't scale good.